cli/src/merge_tools/external.rs

```diff
@@ -308,10 +313,53 @@ pub fn run_mergetool_external(
         }),
         Err(new_file_ids) => conflict.with_new_file_ids(&new_file_ids),
     };
-    let mut tree_builder = MergedTreeBuilder::new(tree.id());
     tree_builder.set_or_remove(repo_path.to_owned(), new_tree_value);
+    Ok(())
+}
+
+pub fn run_mergetool_external(
+    ui: &Ui,
+    path_converter: &RepoPathUiConverter,
+    editor: &ExternalMergeTool,
+    tree: &MergedTree,
+    merge_tool_files: &[MergeToolFile],
+    default_conflict_marker_style: ConflictMarkerStyle,
+) -> Result<(MergedTreeId, Option<MergeToolPartialResolutionError>), ConflictResolveError> {
+    // TODO: add support for "dir" invocation mode, similar to the
+    // "diff-invocation-mode" config option for diffs
+    let mut tree_builder = MergedTreeBuilder::new(tree.id());
+    let mut partial_resolution_error = None;
+    for (i, merge_tool_file) in merge_tool_files.iter().enumerate() {
+        writeln!(
+            ui.status(),
+            "Resolving conflicts in: {}",
+            path_converter.format_file_path(&merge_tool_file.repo_path)
+        )?;
+        match run_mergetool_external_single_file(
+            editor,
+            tree.store(),
+            merge_tool_file,
+            default_conflict_marker_style,
+            &mut tree_builder,
+        ) {
+            Ok(()) => {}
+            Err(err) if i == 0 => {
+                // If the first resolution fails, just return the error normally
+                return Err(err);
+            }
+            Err(err) => {
+                // Some conflicts were already resolved, so we should return an error with the
+                // partially-resolved tree so that the caller can save the resolved files.
+                partial_resolution_error = Some(MergeToolPartialResolutionError {
+                    source: Box::new(err),
+                    resolved_count: i,
+                });
+                break;
```

Comment:
Not for this PR, but I don't think exiting at the first unresolved conflict provides good UX.

If the merge tool can automatically resolve conflicts, user will probably want to process all resolvable files. If the tool is interactive, user might want to terminate in the middle of the process at arbitrary timing.